### PR TITLE
Fix GPIOTE crashes by checking everything is ok

### LINKED ISSUE #5211: nRF52 RotaryIO: raise error when more than 4 encoders are used
I have a small macropad that has 5 rotary encoders.  It appears that CircuitPython has a hardcoded limit of 4 rotary encoders to be setup on the nRF52840. (I use a BlueMicro840). 

I created a [test program](http://pykey.jpconstantineau.com/docs/testing_hardware/encoders) to track the position of all 5 encoders.  In the first iteration of the program, I initialize 5 encoders in sequence 1,2,3,4,5 and the 5th one doesn't respond.
In the second iteration, I moved the initializer for encoder 5 up to be between 3 and 4 (1,2,3,5,4). In this sequence, it's the 4th one that didn't respond.

It would be nice to be able to configure the maximum number of rotary encoders to be initialized.  I understand that 4 is a reasonable limit for most users but with keyboards and more users looking at HID, I suspect that this will be a parameter that might need to be exposed.  

Could it be added into boot.py and not at CP compile time (i.e. hardcoded)? 

macropad in question: [miniMacro5](https://www.tindie.com/products/leafcutter/minimacro5-custom-programmable-keypad/)

### LINKED ISSUE #5240: nRF52840 hardfaults on countio.deinit()
### CircuitPython version

```python
Adafruit CircuitPython 7.0.0-beta.0 on 2021-08-24; Adafruit Feather nRF52840 Express with nRF52840
```

### Code/REPL

```python
import board
import alarm
import countio

alarm_pin = alarm.pin.PinAlarm(pin=board.A2, value=False)
counter = countio.Counter(board.A3)

try:
    alarm.light_sleep_until_alarms(alarm_pin)
    0 / 0

except Exception as e:
    counter.deinit()
    import traceback
    traceback.print_exception(type(e),e,e.__traceback__)
```

### Behavior

Crashes to hardfault handler when code is run, and alarm pin is triggered.
Not including the traceback module and traceback print will make it not crash immediately , but rather when you try to reload the program, or when loading REPL.

### Description

Connected to PC via USB (so lightsleep is fakesleep)

### Additional information

Without counter.deinit() the program will run multiple times just fine. 
It is not really needed here, but was in my code already as it was ported from an RP2040 which does require a deinit to run the same code again.

### PR DESCRIPTION
**Pull request by tannewt**
Fixes #5240 and fixes #5211

### CODE REVIEW

**Review by dhalbert — DISMISSED** [review]
LGTM; I fast-forwarded through your stream and followed the thinking. It would be great if others could test this on hardware.

**Comment by jpconstantineau** [comment]
I'll be able to check the encoders on my hardware once it lands on S3.

**Comment by microDev1** [comment]
@jpconstantineau You can test the build artifacts from this PR. They are [here](https://github.com/adafruit/circuitpython/actions/runs/1175907562).

File: ports/nrf/common-hal/microcontroller/Processor.c [hunk 1]
```diff
@@ -136,6 +138,12 @@ mcu_reset_reason_t common_hal_mcu_processor_get_reset_reason(void) {
         r = RESET_REASON_WATCHDOG;
     } else if (reset_reason_saved & POWER_RESETREAS_SREQ_Msk) {
         r = RESET_REASON_SOFTWARE;
+        // Our "deep sleep" is still actually light sleep followed by a software
+        // reset. Adding this check here ensures we treat it as-if we're waking
+        // from deep sleep.
+        if (sleepmem_wakeup_event != SLEEPMEM_WAKEUP_BY_NONE) {
```
Comment (microDev1) [inline ports/nrf/common-hal/microcontroller/Processor.c:144]: Build failure with `undefined reference to 'sleepmem_wakeup_event'` on some boards.

Comment (jpconstantineau) [reply ports/nrf/common-hal/microcontroller/Processor.c:144]: The BlueMicro840 built fine. Tested it with 5 and 4 encoders.
5th initialization fails and `RuntimeError: All channels in use` exception comes out as expected.

With 4 encoders only, all 4 work as expected.

Haven't tested with using other GPIOTE consumers but I expect rotaryio will be fine if the above test is good.

**Comment by jpconstantineau** [comment]
The BlueMicro840 built fine. Tested it with 5 and 4 encoders.
5th initialization fails and `RuntimeError: All channels in use` exception comes out as expected.

With 4 encoders only, all 4 work as expected.

Haven't tested with using other GPIOTE consumers but I expect rotaryio will be fine if the above test is good.

**Comment by mrdalgaard** [comment]
Tested that countio.deinit() works (does not crash), and alarm + countio can run concurrently

**Comment by mrdalgaard** [comment]
It is however possible to initialize both pin_alarm and countio on the same pin if done in specific order

```
alarm_pin = alarm.pin.PinAlarm(pin=board.A2, value=False)
counter = countio.Counter(board.A2)

```
Will run, but alarm will not trigger on light sleep (it will on deep sleep)

```
counter = countio.Counter(board.A2)
alarm_pin = alarm.pin.PinAlarm(pin=board.A2, value=False)

```
Will error with:
```

Traceback (most recent call last):
  File "code.py", line 7, in <module>
ValueError: A2 in use
```

**Review by dhalbert — APPROVED** [review]
:+1:

**Review by microDev1 — APPROVED** [review]

**Pull request merged**